Fix word completion in VS2019

jaredpar opened the pull request:
Visual Studio 2019 introduces a new completion API that is asynchronous. VsVim must support this async completion in order to implement the word completion behavior (`ctrl-n`, `ctrl-p`). 

Unfortunately Visual Studio supports both async and legacy completion dependent on the content type of the `ITextBuffer`. This is fine for most language services as they are specific to a content type and hence support legacy or async. VsVim though can do word completion in any content hence must support both. 

The actual async support is not a lot of work and is fairly straight forward. In order to support both in Visual Studio, VimApp and our unit tests though I had to do a lot of work in how the code is built. Most significantly introducing a shared source project that contains completion that can be linked into all those scenarios.

The legacy completion will show up as different here due to the file moves. But the changes are really just cosmetic: moving to a new namespace and removing an interface. 

Happy to go into more detail here about any of the changes. This took much longer than I thought it would 😦 

closes #2466

jaredpar commented:
Testing is tricky here. This is probably the first change that really has a large difference between different Visual Studio versions that intersects with our test code. Locally I had everything passing on my VS2019 instance but AppVeyor is roughly testing the VS2017 experience. 

Think this reveals a bit of a test gap. Need to be testing both experiences here. The build infra is setup to do this but we don't exercise it in CI.

I'd been mulling over the idea of switching over to Azure DevOps anyways. Most of the .NET repositories haev been moving there. Doing both runs in parallel there would be pretty trivial. May have to look into that later.

ricksladkey commented on Src/VimCompletion/Implementation/WordCompletion/Async/WordAsyncCompletionSession.cs:
```diff
@@ -0,0 +1,128 @@
+            _asyncCompletionSession = asyncCompletionSession;
+            _asyncCompletionSession.Dismissed += delegate { OnDismissed(); };
+            _vsTextView = vsEditorAdaptersFactoryService?.GetViewAdapter(_textView);
+            if (_vsTextView is object)
```
World's best non-null check!

ricksladkey commented on Src/VimCompletion/Implementation/WordCompletion/Async/WordAsyncCompletionSession.cs:
```diff
@@ -0,0 +1,128 @@
+        /// is unfortunate for VsVim as control is held down for the duration of a completion session. This ... method
+        /// is used to reset the opacity to 1.0 
+        /// </summary>
+        private void ResetTipOpacity(object sender, EventArgs e)
```
Does this mean that the opacity fades a little over the course of 250 msec in a triangle wave pattern? Is it noticeable as a subtle "pulsing"? I will build and test the branch to see if I can notice it. This is different than the way that the old completion session worked around the opacity problem.

ricksladkey replied:
I'm thinking the opacity does this:

```
\    |\    |\
 \   | \   | \
  \  |  \  |  \
   \ |   \ |   \
    \|    \|    \
```

jaredpar replied:
In general it should not do this. The implementation of `SetOpacity` stops the `DispatchTimer` they are using to implement the fade out effect. The reason we're running this on a timer though is that every time a VsVim developer presses control, say if they  take hands off keyboard and then hit ctrl-n again, then the timer will start anew. 

So the subsequent calls to `SetOpacity` should be no-ops in the majority case.

jaredpar replied:
Note: the specific method here is `SimpleTextViewWindow.SetOpacity` and it's in the MS.VS.Editor.Implementation.dll. That should give you a better idea of how this is working under the hood.

ricksladkey commented:
I tried this without success. VsVim is still enabled as an extension but failed to load:

```
  <entry>
    <record>1484</record>
    <time>2019/05/16 23:10:35.175</time>
    <type>Information</type>
    <source>VisualStudio</source>
    <description>Begin package load [VsVimPackage]</description>
    <guid>{A284D12C-1E96-451B-A3B0-5486A1BEB6CA}</guid>
  </entry>
  <entry>
    <record>1485</record>
    <time>2019/05/16 23:10:35.237</time>
    <type>Error</type>
    <source>VisualStudio</source>
    <description>LegacySitePackage failed for package [VsVimPackage]Source: &apos;System.ComponentModel.Composition&apos; Description: No exports were found that match the constraint: &#x000A;&#x0009;ContractName&#x0009;Vim.IVim&#x000A;&#x0009;RequiredTypeIdentity&#x0009;Vim.IVim&#x000D;&#x000A;</description>
    <guid>{A284D12C-1E96-451B-A3B0-5486A1BEB6CA}</guid>
    <hr>80131500</hr>
    <errorinfo></errorinfo>
  </entry>
  <entry>
    <record>1486</record>
    <time>2019/05/16 23:10:35.248</time>
    <type>Error</type>
    <source>VisualStudio</source>
    <description>SetSite failed for package [VsVimPackage](null)</description>
    <guid>{A284D12C-1E96-451B-A3B0-5486A1BEB6CA}</guid>
    <hr>80131500</hr>
    <errorinfo></errorinfo>
  </entry>
  <entry>
    <record>1487</record>
    <time>2019/05/16 23:10:35.260</time>
    <type>Error</type>
    <source>VisualStudio</source>
    <description>End package load [VsVimPackage]</description>
    <guid>{A284D12C-1E96-451B-A3B0-5486A1BEB6CA}</guid>
    <hr>80131500</hr>
    <errorinfo></errorinfo>
  </entry>
```

It's true that I tried to merge it (without conflicts) into my master, so I'll try again just using the PR branch as is.

ricksladkey commented:
I retested again on both VS2017 and VS2019 and got the same symptoms. I'll try running in under the debugger in the experimental instance rather than installing into my normal installations.

```
---------------------------
Microsoft Visual Studio
---------------------------
The 'VsVimPackage' package did not load correctly.



The problem may have been caused by a configuration change or by the installation of another extension. You can get more information by examining the file 'C:\Users\jrs\AppData\Roaming\Microsoft\VisualStudio\15.0_454c4855\ActivityLog.xml'.

Restarting Visual Studio could help resolve this issue.

Continue to show this error message?
---------------------------
Yes   No   
---------------------------
```

ricksladkey commented:
Under the debugger, using VS2019, both VimApp and Visual Studio (experimental instance) fail.

VimApp exception:

```
System.ComponentModel.Composition.ImportCardinalityMismatchException
  HResult=0x80131500
  Message=No exports were found that match the constraint: 
	ContractName	Vim.IWordCompletionSessionFactoryService
	RequiredTypeIdentity	Vim.IWordCompletionSessionFactoryService
  Source=System.ComponentModel.Composition
  StackTrace:
   at System.ComponentModel.Composition.Hosting.ExportProvider.GetExports(ImportDefinition definition, AtomicComposition atomicComposition)
```

Visual Studio exception:

```
System.ComponentModel.Composition.ImportCardinalityMismatchException
  HResult=0x80131500
  Message=No exports were found that match the constraint: 
	ContractName	Vim.IVim
	RequiredTypeIdentity	Vim.IVim
  Source=System.ComponentModel.Composition
  StackTrace:
   at System.ComponentModel.Composition.Hosting.ExportProvider.GetExports(ImportDefinition definition, AtomicComposition atomicComposition)
```

The full stack trace isn't much more interesting:

```
 	System.ComponentModel.Composition.dll!System.ComponentModel.Composition.Hosting.ExportProvider.GetExports(System.ComponentModel.Composition.Primitives.ImportDefinition definition, System.ComponentModel.Composition.Hosting.AtomicComposition atomicComposition)	Unknown
 	System.ComponentModel.Composition.dll!System.ComponentModel.Composition.Hosting.ExportProvider.GetExportsCore(System.Type type, System.Type metadataViewType, string contractName, System.ComponentModel.Composition.Primitives.ImportCardinality cardinality)	Unknown
 	System.ComponentModel.Composition.dll!System.ComponentModel.Composition.Hosting.ExportProvider.GetExportedValueCore<Vim.IVim>(string contractName, System.ComponentModel.Composition.Primitives.ImportCardinality cardinality)	Unknown
 	System.ComponentModel.Composition.dll!System.ComponentModel.Composition.Hosting.ExportProvider.GetExportedValue<Vim.IVim>(string contractName)	Unknown
 	System.ComponentModel.Composition.dll!System.ComponentModel.Composition.Hosting.ExportProvider.GetExportedValue<Vim.IVim>()	Unknown
>	VsVim.dll!Vim.VisualStudio.VsVimPackage.Initialize() Line 47	C#
 	Microsoft.VisualStudio.Shell.11.0.dll!Microsoft.VisualStudio.Shell.Package.Microsoft.VisualStudio.Shell.Interop.IVsPackage.SetSite(Microsoft.VisualStudio.OLE.Interop.IServiceProvider sp)	Unknown
 	[Native to Managed Transition]	
 	[Managed to Native Transition]	
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.DataSource.Invoke(string verb, object pvaIn, out object pvaOut)	Unknown
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.PlatformUI.VsCommand.Execute(object parameter)	Unknown
 	PresentationFramework.dll!MS.Internal.Commands.CommandHelpers.CriticalExecuteCommandSource(System.Windows.Input.ICommandSource commandSource, bool userInitiated)	Unknown
 	PresentationFramework.dll!System.Windows.Controls.MenuItem.InvokeClickAfterRender(object arg)	Unknown
 	WindowsBase.dll!System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate callback, object args, int numArgs)	Unknown
 	WindowsBase.dll!System.Windows.Threading.ExceptionWrapper.TryCatchWhen(object source, System.Delegate callback, object args, int numArgs, System.Delegate catchHandler)	Unknown
 	WindowsBase.dll!System.Windows.Threading.DispatcherOperation.InvokeImpl()	Unknown
 	WindowsBase.dll!System.Windows.Threading.DispatcherOperation.InvokeInSecurityContext(object state)	Unknown
 	WindowsBase.dll!MS.Internal.CulturePreservingExecutionContext.CallbackWrapper(object obj)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state)	Unknown
 	WindowsBase.dll!MS.Internal.CulturePreservingExecutionContext.Run(MS.Internal.CulturePreservingExecutionContext executionContext, System.Threading.ContextCallback callback, object state)	Unknown
 	WindowsBase.dll!System.Windows.Threading.DispatcherOperation.Invoke()	Unknown
 	WindowsBase.dll!System.Windows.Threading.Dispatcher.ProcessQueue()	Unknown
 	WindowsBase.dll!System.Windows.Threading.Dispatcher.WndProcHook(System.IntPtr hwnd, int msg, System.IntPtr wParam, System.IntPtr lParam, ref bool handled)	Unknown
 	WindowsBase.dll!MS.Win32.HwndWrapper.WndProc(System.IntPtr hwnd, int msg, System.IntPtr wParam, System.IntPtr lParam, ref bool handled)	Unknown
 	WindowsBase.dll!MS.Win32.HwndSubclass.DispatcherCallbackOperation(object o)	Unknown
 	WindowsBase.dll!System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate callback, object args, int numArgs)	Unknown
 	WindowsBase.dll!System.Windows.Threading.ExceptionWrapper.TryCatchWhen(object source, System.Delegate callback, object args, int numArgs, System.Delegate catchHandler)	Unknown
 	WindowsBase.dll!System.Windows.Threading.Dispatcher.LegacyInvokeImpl(System.Windows.Threading.DispatcherPriority priority, System.TimeSpan timeout, System.Delegate method, object args, int numArgs)	Unknown
 	WindowsBase.dll!MS.Win32.HwndSubclass.SubclassWndProc(System.IntPtr hwnd, int msg, System.IntPtr wParam, System.IntPtr lParam)	Unknown
```

ricksladkey commented:
OK, I don't there is anything wrong on my machine except that MEF problems are above my pay grade!

I tested some older commits in this branch and some work and some don't:

- 2937474e6a8bb7c09802003d2afdcb09b1005ecc - works
- 63b9ee63b736e8f1db5e31d50d26dda5094c8594 - doesn't work

I could do some more binary searching to find the problematic commit, but I'm pretty sure you if retest that you will see the same problem and that you will probably know exactly what the problem is.

ricksladkey commented:
One more this, the good news is that the unit tests **do** pass for me on VS2019.

jaredpar commented:
Thanks for finding two commits where it works / doesn't work. That helped me narrow down the problem. Looks like my work to fix all the unit tests ended up breaking the actual app and extension. Pretty sure I understand the fix here.

jaredpar commented:
Do understand the problem and why it was working for me originally here:

- VimApp was working because it had the VimCompletion project referenced, and added the export. I deleted this accidentally as a part of the refactoring I did to get the unit tests passing.
- VS was working because I had temporarily unloaded the Vs2015 and Vs2017 projects to work around a VS editor bug. That meant though when I built and deployed only Vs2019 was deployed and hence a single export of `IWordCompletionSessionFactoryService` was available. Once I re-loaded them there are now three copies being deployed and that breaks the expectation that there is a single export. 

The VimApp problem is pretty simple to fix. Just need to add the reference back and fix up the MEF composition.

The VS problem is a bit harder. There are a few ways that I could approach this: 

1. Move the creation of `IWordCompletionSession` to `IVimHost`. This removes the `[Import]` declaration entirely from VimCore. Then the VS specific instance could be exposed via `ISharedService`.
2. Move the import of `IWordCompletionSessionFactoryService` to `ImportMany` and add a method to distinguish which one should be used here. 

At the moment I'm leaning towards 1 as that's the way the infrastructure is geared today. It bothers me a bit though that this is really a core editor component, not a VS component, and hence somewhat wrong to put VS services into the mix here to complete the composition.

Going to think about it a bit today and play with the implementations to see which work better.

jaredpar commented:
Just when you think you understand a MEF problem you remember more MEF rules. Need to always keep in mind that MEF is based on `string` instances not `type`. This means that if I export a type named `WordAsyncCompletionsSessionFactory` in two different assemblies then they are identical from the perspective. Hence an `Import(typeof(WordAsyncCompletionSessionFactory))` will fail with a carnality mismatch. 

The fundamental mistake I made with my MEF plan here was thinking of the VS specific assemblies in isolation. Functionally only one is valid at execution time however all three end up in the MEF composition. That means the `Export` for the async service actually happens three times even though only one is valid. That's a problem. Need to move to versioned string exports here.

This got a lot more complicated but at least I think I fully grasp the problem here now. 

Hopefully 😉

jaredpar commented:
I was wrong. The hole gets deeper.

jaredpar commented:
Going to be some pushes today. Please continue to hold off on reviewing. Doing the pushes so I can validate how the tests run in AppVeyor. I will add a comment once I'm actually ready for a review again. 

Good news though: I seem to have gotten to the bottom of the composition issues here. Now tests failing in 2019 because the async implementation is not actually compliant to the vim model.

jaredpar commented:
This change is ready for review again. Some notes about the final state:

1. This does involve a bit of MEF wizardry but this as simple as I can think to make it given the scenarios that we support. I added a couple of README.md documents to explain the logic here, and some basic do and don't. **More than happy** to expand on that documentation where it's not clear. Took me many hours to figure this all out, want to make sure it's documented fairly well. The scenarios in question are:
    1. Testing word completion in our unit tests
    1. Using word completion in VimApp
    1. Using word completion inside Visual Studio (end customer experience)
1. There is now a significant implementation difference between the various Visual Studio versions that VsVim supports yet we're only verifying one of them in PRs. That introduces risk of regression. This has always been there but the behavior differences have been so small the risk was negligable. Now the risk seems fairly real. Need to change our CI so that we validate our VS2019 and VS2017 / VS2015 behavior at the same time. Mentioned this earlier but feel more strongly about it now.
1. There is a behavior difference between the legacy and async implementations of word completion. The async completion APIs don't have the raw fidelity of the legacy ones hence the vim behavior can't be implemented exactly the same way. Specifically in vim the buffer updates when an item is selected in the completion while in async completion it only updates after it's committed. I've looked around a bit and just don't see an obvious way to make this work with the async APIs. But overall word completion does work again so it's a step forward but hopefully I can think of a way to close this gap. 

Sorry for the big change here. I've tried to simplify it a few times (for my own sanity if nothing else) but there doesn't seem to be an obvious way to do it.

ricksladkey commented on Src/VsSpecific/VsSpecific/README.md:
```diff
@@ -0,0 +1,63 @@
+- Open `%LOCALAPPDATA%\Microsoft\VisualStudio`
+- Open the directory matching `16.*Exp`
+- Open `ComponentModelCache\Microsoft.VisualStudio.Default.err` 
+
```
This explanation helps a lot and this tip will help next time I run into MEF problems.
